src/main/java/jcifs/smb1/util/transport/Transport.java

```diff
@@ -0,0 +1,274 @@
+ * disconnecting, sending, and receiving can be syncronized
+ * properly. Apparatus is provided to send and receive requests
+ * concurrently.
+ */
+
+public abstract class Transport implements Runnable {
+
+    static int id = 0;
+    static LogStream log = LogStream.getInstance();
+
+    public static int readn( InputStream in,
+                byte[] b,
+                int off,
+                int len ) throws IOException {
+        int i = 0, n = -5;
+
+        while (i < len) {
+            n = in.read( b, off + i, len - i );
+            if (n <= 0) {
+                break;
+            }
+            i += n;
+        }
+
+        return i;
+    }
+
+    /* state values
+     * 0 - not connected
+     * 1 - connecting
+     * 2 - run connected
+     * 3 - connected
+     * 4 - error
+     */
+    int state = 0;
+
+    String name = "Transport" + id++;
+    Thread thread;
+    TransportException te;
+
+    protected HashMap response_map = new HashMap( 4 );
+
+    protected abstract void makeKey( Request request ) throws IOException;
+    protected abstract Request peekKey() throws IOException;
+    protected abstract void doSend( Request request ) throws IOException;
+    protected abstract void doRecv( Response response ) throws IOException;
+    protected abstract void doSkip() throws IOException;
+
+    public synchronized void sendrecv( Request request,
+                    Response response,
+                    long timeout ) throws IOException {
+            makeKey( request );
+            response.isReceived = false;
+            try {
+                response_map.put( request, response );
+                doSend( request );
+                response.expiration = System.currentTimeMillis() + timeout;
+                while (!response.isReceived) {
+                    wait( timeout );
+                    timeout = response.expiration - System.currentTimeMillis();
+                    if (timeout <= 0) {
+                        throw new TransportException( name +
+                                " timedout waiting for response to " +
+                                request );
+                    }
+                }
+            } catch( IOException ioe ) {
+                if (log.level > 2)
+                    ioe.printStackTrace( log );
+                try {
+                    disconnect( true );
+                } catch( IOException ioe2 ) {
+                    ioe2.printStackTrace( log );
+                }
+                throw ioe;
+            } catch( InterruptedException ie ) {
+                throw new TransportException( ie );
+            } finally {
+                response_map.remove( request );
+            }
+    }
+    private void loop() {
+        while( thread == Thread.currentThread() ) {
+            try {
+                Request key = peekKey();
+                if (key == null)
+                    throw new IOException( "end of stream" );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Exception [IOException] is explicitly thrown on this line but is caught in the outer catch clause with generic exception on line [112].  Consider adding a catch clause of this specific exception in the outer try statement.